rinad/src/ipcm/ipcm.cc

```diff
@@ -395,6 +400,26 @@ IPCManager_::assign_to_dif(Addon* callee, Promise* promise,
 	IPCPTransState* trans;
 
 	try {
+		if (is_any_ipcp_assigned_to_dif(dif_name)) {
+			ss << "There is already an IPCP assigned to DIF "
+				<< dif_name.toString()
+				<< " in this system.";
+			FLUSH_LOG(ERR, ss);
+			throw rina::AssignToDIFException();
+		}
+
+		ipcp = lookup_ipcp_by_id(ipcp_id, false);
+		if (ipcp->get_type() == rina::NORMAL_IPC_PROCESS) {
+			// Load all the plugins required from by template, but
+			// first release the ipcp lock, to avoid lock-ups.
+			ipcp->rwlock.unlock();
+
+			catalog.load_by_template(callee, ipcp_id, dif_template);
```

Comment:
As I have already said, this pull request is needed for experimentation, that we must start **immediately**.
This is why I want to merge it now - we can start experimentation in parallel with the improvements that you suggest.

It won't be complicated to fix afterwards - it's will just be the same. The other pull requests that are pending are implementation of new policy sets, that don't care at all about the IPCM internal implementation. It does not make sense to stall this for internal rearrangments that we have do anyway - since this catalog thing is basically part of second iteration SDK.